src/coprocessor/dag/rpn_expr/types/expr_builder.rs

```diff
@@ -278,6 +296,23 @@ where
     // Map pb func to `RpnFnMeta`.
     let func = fn_mapper(tree_node.get_sig(), tree_node.get_children())?;
     let args = tree_node.take_children().into_vec();
+
+    // Only Int/Real/Duration/Decimal/Bytes/Json will be decoded
+    let datums = datum::decode(&mut tree_node.get_val())?;
+    let implicit_args = datums
+        .into_iter()
+        .map(|d| match d {
+            Datum::I64(n) => ScalarValue::Int(Some(n)),
+            Datum::U64(n) => ScalarValue::Int(Some(n as i64)),
+            Datum::F64(n) => ScalarValue::Real(Some(Real::new(n).unwrap())),
+            Datum::Dur(dur) => ScalarValue::Duration(Some(dur)),
+            Datum::Bytes(bytes) => ScalarValue::Bytes(Some(bytes)),
+            Datum::Dec(dec) => ScalarValue::Decimal(Some(dec)),
+            Datum::Json(json) => ScalarValue::Json(Some(json)),
+            _ => unreachable!(),
```

Comment:
- This is not unreachable and may lead to panic when TiDB passes unexpected data.
- We may need to re-consider the implementation when we need a NULL.

Reply:
There is no panic because unexpected data will cause early return in L301.

Reply:
It can be Datum::Null

Reply:
I think we should redesign the `ScalarValue`, like:
```rust
enum ScalarValue {
	None,
	Int(Int),
	Real(Real),
	...
}
```
The world will be better.

Reply:
Perhaps a good idea? But logically, I think a `ScalarValue` should be able to distinguish from Null of different types.

Reply:
I don't think `Null` should be associated with any type. Null is null, is nothing. Should we distinguish the different of `nothing`?

Reply:
Is there possibly a function, which outputs Null given two `Null`s of the same type while throws an error given two different types? I'm not quite sure about it.

Reply:
`ScalarValue` was not designed in this way simply because RPN function accepts `&Option<T>` as arguments. `VectorValue` stores `Vec<Option<T>>` so that `&Option<T>` is very easy and cheap to be built. If `ScalarValue` is defined as `enum ScalarValue { None, Int(Int), ... } `, then there is no cheap way to get a `&Option<T>` from `&ScalarValue`.

In order to make your propose work, we need to change RPN function to accept `Option<&T>`. However this introduces extra cost for `VectorValue` which is `Vec<Option<T>>`, since `Option<&T>` is not a simple reference of `&Option<T>`. Although this is a simple `&Option<T>::as_ref() -> Option<&T>` function call, there is indeed extra cost.

Reply:
It's OK to use the current implementation of `ScalarValue`.